### PR TITLE
[common] cpu/gpu mix trace fix

### PR DESCRIPTION
**Pull request by super-dainiu**
#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

**Comment by super-dainiu** [comment]
hmm the test is weird...

File: test/algo/compression/pruning/test_concrete_trace_mmdetection.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
     # 'ddod/ddod_r50_fpn_1x_coco',
     # 'deepfashion/mask-rcnn_r50_fpn_15e_deepfashion',
     # 'deformable_detr/deformable-detr_r50_16xb2-50e_coco',
-    'detr/detr_r18_8xb2-500e_coco',
+    # 'detr/detr_r18_8xb2-500e_coco',
```
Comment (super-dainiu) [inline test/algo/compression/pruning/test_concrete_trace_mmdetection.py:28]: these two nets would give weird output if I applied to `graph.eliminate_dead_code()`.

File: nni/common/concrete_trace_utils/concrete_tracer.py [hunk 2]
```diff
@@ -214,7 +216,7 @@ class ConcreteTracer(TracerBase):
     node_to_originating_module : Dict[torch.fx.Node, str] = {}
 
     @compatibility(is_backward_compatible=True)
-    def __init__(self):
+    def __init__(self, cpu_offload = False):
```
Comment (J-shang) [inline nni/common/concrete_trace_utils/concrete_tracer.py:219]: please add a param docstring for `cpu_offload`

File: nni/common/concrete_trace_utils/concrete_tracer.py [hunk 3]
```diff
@@ -275,36 +278,63 @@ def run_target(self, kind: str, target: Target, args: Tuple[Any, ...], kwargs: D
...
+                assert isinstance(target, str)
+                mod = self.fetch_attr(target)
+                if self.cpu_offload:
+                    mod.cuda()  # how it works in ddp?
+                if _orig_getattr(mod, '__module__', None) != 'nni.common.concrete_trace_utils.concrete_tracer' \
+                    and hasattr(mod, '__globals__'):
+                    _autowrap_check(self, mod.__globals__, self._autowrap_function_ids, self.autowrap_leaf_pairs, self.agfunc_dict)
+                result = OperatorPatcherContext.patch_run(mod, *args, **kwargs)
+                if self.cpu_offload:
+                    mod.cpu()
```
Comment (J-shang) [inline nni/common/concrete_trace_utils/concrete_tracer.py:318]: should we check if `mod` is on cpu before? Is it possible that `mod` was originally on the gpu?

File: nni/common/concrete_trace_utils/concrete_tracer.py [hunk 4]
```diff
@@ -275,36 +278,63 @@ def run_target(self, kind: str, target: Target, args: Tuple[Any, ...], kwargs: D
...
+            else:
+                raise RuntimeError()
+            return result
+
+        with self.do_temp_disable(call=True):
+            result = run(kind, target, args, kwargs)
+            if self.cpu_offload:
+                if isinstance(result, torch.Tensor):
+                    result = result.cpu()
+                elif isinstance(result, (list, dict, tuple)):
```
Comment (J-shang) [inline nni/common/concrete_trace_utils/concrete_tracer.py:331]: a little concern is could a module result be a customized data structure? i.e., when I take `BertModel` as a leaf module and it return an output like `BertOuput`?

File: nni/common/concrete_trace_utils/concrete_tracer.py [hunk 5]
```diff
@@ -542,15 +572,15 @@ def proxy_placeholder(name: str):
...
-              autowrap_modules: Tuple[str] | None = None,
-              autowrap_leaf_function = None,
-              autowrap_leaf_class = None,
-              leaf_module = None,
-              fake_middle_class = None,
-              concrete_args: Union[Dict[str, Any], Tuple],
-              use_operator_patch: bool = True,
-              operator_patch_backlist: List[str] | None = None,
-              forward_function_name: str = 'forward') -> Graph:
+            autowrap_modules: Tuple[str] | None = None,
```
Comment (J-shang) [inline nni/common/concrete_trace_utils/concrete_tracer.py:575]: NNI prefer alignment with brackets.

File: nni/common/concrete_trace_utils/concrete_tracer.py [hunk 6]
```diff
@@ -992,6 +1022,10 @@ def getattr_wrapper(obj, *args):
             pass
 
         self.submodule_paths = None
+        with MagicMethodPatcher():
+            name = root.__class__.__name__ if isinstance(root, torch.nn.Module) else root.__name__
```
Comment (J-shang) [inline nni/common/concrete_trace_utils/concrete_tracer.py:1026]: what's this name used for?

File: nni/common/concrete_trace_utils/concrete_tracer.py (hunk 3 above)
Comment (super-dainiu) [reply nni/common/concrete_trace_utils/concrete_tracer.py:318]: cpu_offload means we should offload to cpu, so `mod`  is assumed to be on GPU

File: nni/common/concrete_trace_utils/concrete_tracer.py (hunk 5 above)
Comment (super-dainiu) [reply nni/common/concrete_trace_utils/concrete_tracer.py:575]: i see

File: nni/common/concrete_trace_utils/concrete_tracer.py (hunk 6 above)
Comment (super-dainiu) [reply nni/common/concrete_trace_utils/concrete_tracer.py]: fixed

File: nni/common/concrete_trace_utils/concrete_tracer.py (hunk 3 above)
Comment (J-shang) [reply nni/common/concrete_trace_utils/concrete_tracer.py:318]: So if `cpu_offload == True`, should we give a warning if the original model has some params on GPU?

**Review by J-shang — APPROVED** [review]

File: nni/common/concrete_trace_utils/concrete_tracer.py (hunk 2 above)
Comment (QuanluZhang) [reply nni/common/concrete_trace_utils/concrete_tracer.py:219]: not added yet?

Comment (super-dainiu) [reply nni/common/concrete_trace_utils/concrete_tracer.py:219]: i add it in the function doc of concrete_trace

File: test/pytest.ini [hunk 7]
```diff
@@ -1,13 +0,0 @@
-[pytest]
```
Comment (QuanluZhang) [inline test/pytest.ini:1]: what is this file used for? why should we delete it?

File: nni/common/concrete_trace_utils/concrete_tracer.py (hunk 2 above)
Comment (QuanluZhang) [reply nni/common/concrete_trace_utils/concrete_tracer.py:219]: got it

File: nni/common/concrete_trace_utils/concrete_tracer.py [hunk 8]
```diff
@@ -275,36 +278,63 @@ def run_target(self, kind: str, target: Target, args: Tuple[Any, ...], kwargs: D
...
+                raise RuntimeError()
+            return result
+
+        with self.do_temp_disable(call=True):
+            result = run(kind, target, args, kwargs)
+            if self.cpu_offload:
+                if isinstance(result, torch.Tensor):
+                    result = result.cpu()
+                elif isinstance(result, (list, dict, tuple)):
+                    result = tree_map(to_cpu, result)
```
Comment (QuanluZhang) [inline nni/common/concrete_trace_utils/concrete_tracer.py:332]: should we output warning message when `result` is not torch.Tensor, list, dict, or tuple?

File: nni/common/concrete_trace_utils/concrete_tracer.py (hunk 3 above)
Comment (super-dainiu) [reply nni/common/concrete_trace_utils/concrete_tracer.py:318]: I think this cannot happen in general cases. The model is always on cpu before tracing, unless user purposely moves it to gpu.

Comment (super-dainiu) [reply nni/common/concrete_trace_utils/concrete_tracer.py:318]: And if gpu can hold the model immediately, there is no need for cpu_offload.

File: nni/common/concrete_trace_utils/concrete_tracer.py (hunk 8 above)
Comment (super-dainiu) [reply nni/common/concrete_trace_utils/concrete_tracer.py:332]: If the result is int, it should not be a warning I guess. But the customized structure is really hard to handle.

Comment (super-dainiu) [reply nni/common/concrete_trace_utils/concrete_tracer.py:332]: Maybe I can add a whitelist for the types

Comment (QuanluZhang) [reply nni/common/concrete_trace_utils/concrete_tracer.py:332]: maybe simply output warning is enough in the `else` branch

Comment (super-dainiu) [reply nni/common/concrete_trace_utils/concrete_tracer.py:332]: done

**Review by QuanluZhang — APPROVED** [review]

**Pull request merged**